spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/ChannelBindingProperties.java

```diff
@@ -84,8 +113,120 @@ else if (binding instanceof Map) {
+				if (bindingPath != null) {
+					return Integer.parseInt(bindingPath.toString());
+				}
+			} catch (NumberFormatException e) {
+				// ignore and just return 1
+			}
+		}
+		return 1;
+	}
+
+	public boolean isPartitionedConsumer(String channelName) {
+		Object binding = bindings.get(channelName);
+		// if the setting is just a target shortcut
+		if (binding == null || binding instanceof String) {
+			return false;
+		}
+		else if (binding instanceof Map) {
+			Map<?, ?> bindingProperties = (Map<?, ?>) binding;
+			Object bindingPath = bindingProperties.get(PARTITIONED);
+			if (bindingPath != null) {
+				return Boolean.valueOf(bindingPath.toString());
+			}
+		}
+		// just return the channel name if not found
+		return false;
+	}
+
+	public boolean isPartitionedProducer(String channelName) {
+		Object binding = bindings.get(channelName);
+		// if the setting is just a target shortcut
+		if (binding == null || binding instanceof String) {
+			return false;
+		}
+		else if (binding instanceof Map) {
+			Map<?, ?> bindingProperties = (Map<?, ?>) binding;
+			return bindingProperties.get(BinderProperties.PARTITION_KEY_EXPRESSION) != null
+					|| bindingProperties.get(BinderProperties.PARTITION_KEY_EXTRACTOR_CLASS) != null;
+		}
+		return false;
+	}
+
+	/**
+	 * Merge general properties provided by 'spring.cloud.stream.consumerProperties.*' with individual binding
+	 * properties supplied via binders.
+	 *
+	 * @param inputChannelName the input channel name
+	 * @return merged consumer properties
+	 */
+	public Properties getConsumerProperties(String inputChannelName) {
+		if (isPartitionedConsumer(inputChannelName)) {
+			Properties channelConsumerProperties = new Properties();
+			if (consumerProperties != null) {
+				channelConsumerProperties.putAll(consumerProperties);
+			}
+			channelConsumerProperties.setProperty(BinderProperties.COUNT,
+					Integer.toString(getInstanceCount()));
+			channelConsumerProperties.setProperty(BinderProperties.PARTITION_INDEX,
+					Integer.toString(getInstanceIndex()));
+			return channelConsumerProperties;
+		}
+		else {
+			return getConsumerProperties();
+		}
+	}
+
+	/**
+	 * Merge general properties provided by 'spring.cloud.stream.producerProperties.*' with individual binding
+	 * properties supplied via binders.
+	 *
+	 * @param outputChannelName the output channel name
+	 * @return merged producer properties
+	 */
+	public Properties getProducerProperties(String outputChannelName) {
+		if (isPartitionedProducer(outputChannelName)) {
+			Properties channelProducerProperties = new Properties();
+			if (this.producerProperties != null) {
+				channelProducerProperties.putAll(this.producerProperties);
+			}
+			channelProducerProperties.put(BinderProperties.MIN_PARTITION_COUNT,
+					Integer.toString(getPartitionCount(outputChannelName)));
+			channelProducerProperties.setProperty(BinderProperties.NEXT_MODULE_COUNT,
+					Integer.toString(getPartitionCount(outputChannelName)));
```

Comment:
Nitpicky: we can probably avoid calling `getPartitionCount(String ChannelName)` twice as it is same for both `minPartitionCount` and `nextModuleCount`.